cirq-core/cirq/transformers/stratify.py

```diff
@@ -240,3 +320,225 @@ def _get_op_class(op: 'cirq.Operation', classifiers: Sequence[Classifier]) -> in
+
+    def add(self, op: ops.Operation) -> None:
+        """Add an operation to this stratum.
+
+        WARNING: For performance reasons, this method does not check whether this stratum can
+            accomodate the given op.  Add operations at your own peril!
+        """
+        self._ops.append(op)
+        self._qubits |= set(op.qubits)
+
+
+class _Strata:
+    """A data structure to organize a collection of strata ('_Stratum's).
+
+    The naming and language in this class imagine that strata are organized into a vertical stack,
+    with time "increasing" as you go "up".  That is, if stratum A precedes stratum B (i.e.,
+    A.time_index < B.time_index), then stratum A is said to be "below" stratum B, and stratum B is
+    said to be "above" stratum A.
+
+    In accordance with this metaphor, we build a '_Strata_ object by adding operations to the stack
+    of strata "from above".
+    """
+
+    def __init__(self, classifiers: Sequence[Classifier]) -> None:
+        self._classifiers = classifiers
+        self._strata: List[_Stratum] = []
+
+        # map from qubit --> the last stratum that adresses that qubit
+        self._qubit_floor: Dict['cirq.Qid', _Stratum] = {}
+
+        # map from a stratum to its index in self._strata
+        self._stratum_index: Dict[_Stratum, int] = {}
+
+    def to_circuit(self) -> circuits.Circuit:
+        return circuits.Circuit(stratum.to_moment() for stratum in self._strata)
+
+    def add(self, op: ops.Operation) -> None:
+        """Add an operation to the lowest stratum possible.
+
+        Strategy:
+        (1) Find the "op_floor" stratum, i.e., the highest stratum that collides with the op.
+        (2) Try to find the lowest stratum that
+            (a) is below the op_floor,
+            (b) can accomodate the op, and
+            (c) can be moved up above the op_floor (without violating causality).
+            If such a "below_stratum" exists, move it above the op_floor add the op to it.
+        (3) If no below_stratum exists, find the lowest stratum above the op_floor that can
+            accomodate the op, and add the op to this "above_stratum".
+        (4) If no above_stratum exists either, add the op to a new stratum above everything.
+        """
+        op_class = _get_op_class(op, self._classifiers)
+        op_floor = self._get_op_floor(op)
+
+        op_is_merged = self._merge_into_below_stratum(op_class, op_floor, op)
+        if not op_is_merged:
+            op_is_merged = self._merge_into_above_stratum(op_class, op_floor, op)
+        if not op_is_merged:
+            self.append_stratum(op_class, op)
+
+    def _get_op_floor(self, op: ops.Operation) -> Optional[_Stratum]:
+        """Get the highest stratum that collides with this op, if there is any."""
+        candidates = [stratum for qubit in op.qubits if (stratum := self._qubit_floor.get(qubit))]
+        return max(candidates, key=lambda stratum: stratum.time_index) if candidates else None
```

Comment:
With regards to handling measurement and control keys, I'll point out that this method is currently doing the work of [`circuits.circuit.get_earliest_accommodating_moment_index`](https://github.com/quantumlib/Cirq/blob/79286a1d31ebe76d92de12647fa6f3ba269d4e27/cirq-core/cirq/circuits/circuit.py#L2732).